sl/adt_op_replace.cc

```diff
@@ -740,7 +740,7 @@ bool tryReplaceCond(
     str << "?cond" << ctx.condVar;
     TGenericVarSet live;
     live.insert(GenericVar(VL_COND_VAR, condVar));
-    AnnotatedInsn *oldInsn = dynamic_cast<AnnotatedInsn *>(locState.insn);
+    AnnotatedInsn *oldInsn = static_cast<AnnotatedInsn *>(locState.insn);
```

Comment:
Although this change could prevent the code from crashing, I do not believe it will make the code actually work.  Why did you use `-args=dump_fixed_point` in the first place?

I think we should change implementation of the option such that it only makes predator dump the fixed point, nothing else.  The fact that the option triggers detection of container shape operations as a side effect must be completely unexpected for users.

Reply:
For printing [symheaps](http://www.fit.vutbr.cz/~isokova/predator/plarena/fp-PL_ArenaAllocate.svg) for every instructions? And yes, it works flawlessly.

Reply:
The code that crashed is not needed for producing these graphs.  `replaceAdtOps()` rewrites control-flow graph of the input program such that low-level pointer manipulation is replaced by high-level operations over abstract data types.  To be clear, are you interested in the red edges in http://www.fit.vutbr.cz/~isokova/predator/plarena/fp-PL_ArenaAllocate.svg ?  Would it work for you if the red edges were black?

Reply:
Honestly, I don't even know what a green box means [http://www.fit.vutbr.cz/~isokova/predator/plarena/fp-PL_ArenaAllocate-loc0052-sh0000-0000.svg](http://www.fit.vutbr.cz/~isokova/predator/plarena/fp-PL_ArenaAllocate-loc0052-sh0000-0000.svg) I guess, yes, it will work for me. Just for information, what do these two things (red arrows, green boxes) mean?

Reply:
Thanks for confirmation!  I will open a separate pull request to make this code disabled in predator unless a user explicitly asks for it.  The green boxes in heap graphs highlight detected container shapes and red edges in trace graphs highlight detected container shape mapping.  The term container shape is defined in the following paper: https://link.springer.com/chapter/10.1007%2F978-3-662-49122-5_21 (also available [here](https://arxiv.org/pdf/1510.07995) for non-students).

Reply:
Good news.  I discovered that we already had an option for this.  So the fix will be trivial in the end.  Could you please check whether pull request  #50 fixes the crash for you?

Reply:
I will skip this change because it would hide a real problem in the code.  As far as I can see, there is no reason why `dynamic_cast<AnnotatedInsn *>(locState.insn)` would return a different value than `static_cast` under normal circumstances.  If it does, it means that `locState.insn` is already invalid.  So it is better to stop immediately to ease debugging of the cause.

Reply:
In the commit kdudka/predator@abcefe0 was similar problem. `dynamic_cast` caused segfault, but `static_cast` didn't on macos.  I can't tell why, but the output (enabled containers detection) is the same as on Linux, so it works with `static_cast`. It's up to you.

Reply:
That is really strange.  I guess that the implementation of `SymHeap` also stops working for you if you redefine `DCAST` as `dynamic_cast`, doesn't it?  There must be either some undefined behavior in the code of predator, or some nasty bug in the toolchain on Mac.  I submitted pull request #51 to put the workaround back in a way that does not make debugging of the root cause more difficult.  Is that acceptable for you?